### ISSUE TITLE
Trigger In-Product Acquisition for optional SDK workloads

### ISSUE THREAD

**Issue by dsplaisted** [dotnet/project-system#6080]
When opening a project that requires an optional SDK workload (such as iOS or Android) that isn't installed, the project system should trigger an in-product acquisition experience to allow the user to click through to install the required workload.  The design for this (from the MSBuild / SDK side) is here: https://github.com/dotnet/designs/pull/104

**Comment by dsplaisted** [comment]
FYI @jjmew, I've filed this issue to track the in-product acquisition experience we've discussed a few times.

**Comment by jjmew** [comment]
Triage: @davkean do you have a recommendation who should work on this?

**Comment by jjmew** [comment]
@cartermp We need your help designing this. This is high priority. Please work with @davkean

**Comment by cartermp** [comment]
Yep, happy to do it. I'm not quite clear on if we get access to evaluation results at "in product acquisition time", since my understanding is that the VS system triggers this when it _attempts_ to load a project. Can IPA also be triggered after evaluation? @davkean or @jmarolf would know more.

@dsplaisted The question in your spec is a good one:

> QUESTION: Will it be OK to put this mapping of .NET workload IDs to VS workload/component IDs in the workload manifest packages?  Or do we need to supply it some other way

I'd defer to @jmarolf since he authored most of the .NET manifests and probably has the best idea for where that mapping should live. What VS tells the installer to acquire ultimately depends on how things are factored and what level of granularity (VS Workload, VS Component Group, VS Component) we're expecting with .NET SDK workloads and what state someone's VS might be in w.r.t what they have installed. @dsplaisted are you the owner of that?

Side Note: the terminology here is going to get confusing for everyone involved 😿

**Comment by davkean** [comment]
@cartermp Let's sync up, I have a good idea how to approach based on existing infrastructure we already have in the product.

**Comment by davidwengier** [comment]
Plan:

* Solution-level IProjectRetargetingManager service that waits to be told about problems
* Project-level (configured? unconfigured?) service that hooks up to evaluation on project load
* When evaluation data comes in, above service calls to its providers, and each one checks for a specific problem
* If problems are found, they are reported to the IProjectRetargetingManager
* Once the solution is loaded (???) the IProjectRetargetingManager calls `IVsTrackProjectRetargeting2.CheckSolutionForRetarget` to kick off the retargeting
* Retargeting is handled by a `IProjectRetargetingProvider` that reports the problems found above

**Comment by davidwengier** [comment]
![image](https://user-images.githubusercontent.com/754264/81899337-ed605780-95fd-11ea-8b91-df9fbf534470.png)
